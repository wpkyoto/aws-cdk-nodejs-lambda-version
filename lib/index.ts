import * as cdk from '@aws-cdk/core';
import { Version } from '@aws-cdk/aws-lambda';
import { NodejsFunction, NodejsFunctionProps } from '@aws-cdk/aws-lambda-nodejs';
import { createHash } from 'crypto';
import { readFileSync } from 'fs';


export interface NodejsLambdaVersionProps extends NodejsFunctionProps {
  /**
   * Path to the entry file (JavaScript or TypeScript).
   *
   * If the `NodejsFunction` is defined in `stack.ts` with `my-handler` as id
   * (`new NodejsFunction(this, 'my-handler')`), the construct will look at `stack.my-handler.ts`
   * and `stack.my-handler.js`.
   */
  entry: string;
}

export class NodejsLambdaVersion extends cdk.Construct {
  /**
   * Created Lambda Function
   */
  public readonly function: NodejsFunction

  /**
   * Function version
   */
  public readonly version: Version

  constructor(scope: cdk.Construct, id: string, props: NodejsLambdaVersionProps) {
    super(scope, id);
    this.function = new NodejsFunction(
      this,
      `${id}NodeJSFunction`,
      props
    );

    this.version = this.function.addVersion(
      createHash('sha256')
        .update(readFileSync(props.entry, 'utf-8'))
        .digest('hex')
    );

  }
}
