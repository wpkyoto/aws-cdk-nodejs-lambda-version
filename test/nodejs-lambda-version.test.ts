import { expect as expectCDK, haveResource, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as NodejsLambdaVersion from '../lib/index';

test('SQS Queue Created', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, "TestStack");
    // WHEN
    new NodejsLambdaVersion.NodejsLambdaVersion(stack, 'MyTestConstruct');
    // THEN
    expectCDK(stack).to(haveResource("AWS::SQS::Queue"));
});

test('SNS Topic Created', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  // WHEN
  new NodejsLambdaVersion.NodejsLambdaVersion(stack, 'MyTestConstruct');
  // THEN
  expectCDK(stack).to(haveResource("AWS::SNS::Topic"));
});
