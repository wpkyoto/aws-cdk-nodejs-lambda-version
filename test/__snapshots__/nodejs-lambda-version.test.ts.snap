// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32ArtifactHash3FE485FF": Object {
      "Description": "Artifact hash for asset \\"cdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32\\"",
      "Type": "String",
    },
    "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32S3Bucket42F24EDA": Object {
      "Description": "S3 bucket for asset \\"cdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32\\"",
      "Type": "String",
    },
    "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32S3VersionKey36395354": Object {
      "Description": "S3 key for asset version \\"cdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructMyTestConstructNodeJSFunction700F25A7": Object {
      "DependsOn": Array [
        "MyTestConstructMyTestConstructNodeJSFunctionServiceRoleBAB50534",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32S3Bucket42F24EDA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32S3VersionKey36395354",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscdbf60bb1de1ec6027f8097f8a3439c94e00cdeccfd5c8d000764c8840d1dd32S3VersionKey36395354",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "test",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructMyTestConstructNodeJSFunctionServiceRoleBAB50534",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructMyTestConstructNodeJSFunctionServiceRoleBAB50534": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructMyTestConstructNodeJSFunctionVersion062ee3e8f4a8046627161501ac2d59558bf42479f85b2065d65b464bddf62221AD6E26FB": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructMyTestConstructNodeJSFunction700F25A7",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
  },
}
`;
